src/Build/Graph/ProjectGraph.cs

```diff
@@ -219,47 +218,113 @@ public ProjectGraph(string entryProjectFile)
+                            var projectsInCycle = new List<string> {projectReferenceConfigurationMetadata.ProjectFullPath};
+                            return (false, projectsInCycle);
+                        }
+                    }
+                }
+                else
+                {
+                    // a new project that has to be evaluated
+                    var loadReference = ProcessNode(projectReferenceConfigurationMetadata, nodeState, projectCollection,
+                        globalProperties);
+                    if (!loadReference.success)
+                    {
+                        if (loadReference.projectsInCycle[0].Equals(parsedProject.Project.FullPath))
+                        {
+                            // we have reached the nth project in the cycle, form error message and throw
+                            loadReference.projectsInCycle.Add(projectReferenceConfigurationMetadata.ProjectFullPath);
+                            loadReference.projectsInCycle.Add(parsedProject.Project.FullPath);
+                            var errorMessage = FormatCircularDependencyError(loadReference.projectsInCycle);
+                            throw new CircularDependencyException(string.Format(
+                                ResourceUtilities.GetResourceString("CircularDependencyInProjectGraph"),
+                                errorMessage));
+                        }
+                        else
+                        {
+                            loadReference.projectsInCycle.Add(projectReferenceConfigurationMetadata.ProjectFullPath);
+                            return (false, loadReference.projectsInCycle);
                         }
-
-                        string projectReferenceFullPath = projectReferenceToParse.GetMetadataValue(FullPathMetadataName);
-
-                        PropertyDictionary<ProjectPropertyInstance> projectReferenceGlobalProperties = GetProjectReferenceGlobalProperties(projectReferenceToParse, globalProperties);
-                        var configurationMetadata = new ConfigurationMetadata(projectReferenceFullPath, projectReferenceGlobalProperties);
-                        projectsToEvaluate.Enqueue(new KeyValuePair<ConfigurationMetadata, ProjectGraphNode>(configurationMetadata, parsedProject));
                     }
                 }
 
-                parentNode?.AddProjectReference(parsedProject);
+                parsedProject.AddProjectReference(_allParsedProjects[projectReferenceConfigurationMetadata]);
+            }
+
+            nodeState[projectToEvaluate] = NodeState.Processed;
+            return (true, null);
+        }
+
+        internal static string FormatCircularDependencyError(List<string> projectsInCycle)
+        {
+            var errorMessage = new StringBuilder(500);
+            errorMessage.AppendLine();
+            for (int i = projectsInCycle.Count - 1; i >= 0; i--)
```

Comment:
It makes me so sad that they only added `StringBuilder.AppendJoin` in .NET Core 2. 😕